lib/note-content-editor.jsx

```diff
@@ -63,6 +93,144 @@ export default class NoteContentEditor extends React.Component {
 		invoke( this, 'editor.focus' );
 	}
 
+	handleReturn = () => {
+		let editorState = this.state.editorState;
+		const key = editorState.getSelection().getFocusKey();
+		const block = editorState.getCurrentContent().getBlockForKey( key );
+		const line = block.getText();
+
+		// matches lines that start with `- ` or `* `
+		// preceded by 0 or more tab characters
+		const listItemRe = /^\t*[-\*]\s/;
+		let listItemMatch;
+
+		if ( line.trim() === '-' || line.trim() === '*' ) {
+			// if current line is just a bullet point and whitesapce,
+			// then clear the line (finishing a list)
+
+			// remove `- ` from the current line
+			editorState = EditorStateModifier.removeRange(
+				editorState,
+				SelectionState.createEmpty( key ).merge( {
+					focusOffset: block.getLength()
+				} )
+			)
+
+			// move selection to the start of the line
+			editorState = EditorState.forceSelection(
+				editorState,
+				SelectionState.createEmpty( key )
+			);
+
+			this.handleEditorStateChange( editorState );
+
+			return 'handled';
+		} else if ( ( listItemMatch = line.match( listItemRe ) ) ) {
```

Comment:
Instead of carrying the conditional through with indentation, why don't we create a new function to handle the behavior which is here in the block…

```js
const trimmedLine = line.trim();
if ( trimmedLine === '-' || trimmedLine === '*' ) {
	return removeBulletAndEndList( editorState, block );
}

const listItemMatch = line.match( listItemRe );
if ( listItemMatch ) {
	const [ bullet ] = listItemMatch;

	return addNewListItem( editorState, bullet );
}
```